sql/parser/type_check.go

```diff
@@ -0,0 +1,432 @@
+
+	case *NullCheck:
+		if _, err := TypeCheckExpr(t.Expr); err != nil {
+			return nil, err
+		}
+		return DummyBool, nil
+
+	case *ExistsExpr:
+		// The subquery within the exists should have been executed before
+		// expression evaluation and the exists nodes replaced with the result.
+
+	case BytesVal, StrVal:
+		return DummyString, nil
+
+	case IntVal:
+		return DummyInt, nil
+
+	case NumVal:
+		return DummyFloat, nil
+
+	case BoolVal:
+		return DummyBool, nil
+
+	case ValArg:
+		// Placeholders should have been replaced before expression evaluation.
+
+	case *QualifiedName:
+		return nil, fmt.Errorf("qualified name \"%s\" not found", t)
+
+	case Tuple:
+		tuple := make(DTuple, 0, len(t))
+		for _, v := range t {
+			d, err := TypeCheckExpr(v)
+			if err != nil {
+				return nil, err
+			}
+			tuple = append(tuple, d)
+		}
+		return tuple, nil
+
+	case DReference:
+		return t.Datum(), nil
+
+	case Datum:
+		return t, nil
+
+	case *Subquery:
+		// The subquery should have been executed before expression evaluation and
+		// the result placed into the expression tree.
+
+	case *BinaryExpr:
+		return typeCheckBinaryExpr(t)
+
+	case *UnaryExpr:
+		return typeCheckUnaryExpr(t)
+
+	case *FuncExpr:
+		return typeCheckFuncExpr(t)
+
+	case *CaseExpr:
+		return typeCheckCaseExpr(t)
+
+	case *CastExpr:
+		return typeCheckCastExpr(t)
+
+	default:
+		return nil, util.Errorf("typeCheck: unsupported expression: %T", expr)
+	}
+
+	return nil, util.Errorf("typeCheck: unexpected expression: %T", expr)
+}
+
+func typeCheckBooleanExprs(exprs ...Expr) (Datum, error) {
+	for _, expr := range exprs {
+		dummyExpr, err := TypeCheckExpr(expr)
+		if err != nil {
+			return nil, err
+		}
+		if _, ok := dummyExpr.(DBool); !ok {
+			return nil, fmt.Errorf("incompatible AND argument type %s", dummyExpr.Type())
+		}
+	}
+	return DummyBool, nil
+}
+
+// TODO(tamird): reduce duplication with evalComparisonExpr.
+func typeCheckComparisonExpr(expr *ComparisonExpr) (Datum, error) {
+	leftType, err := TypeCheckExpr(expr.Left)
+	if err != nil {
+		return nil, err
+	}
+	rightType, err := TypeCheckExpr(expr.Right)
+	if err != nil {
+		return nil, err
+	}
+
+	return typeCheckComparisonOp(expr.Operator, leftType, rightType)
+}
+
+func typeCheckComparisonOp(op ComparisonOp, dummyLeft, dummyRight Datum) (Datum, error) {
+	switch op {
+	case NE:
+		// NE(left, right) is implemented as !EQ(left, right).
+		op = EQ
+	case GT:
+		// GT(left, right) is implemented as LT(right, left)
+		op = LT
+		dummyLeft, dummyRight = dummyRight, dummyLeft
+	case GE:
+		// GE(left, right) is implemented as LE(right, left)
+		op = LE
+		dummyLeft, dummyRight = dummyRight, dummyLeft
+	case NotIn:
+		// NotIn(left, right) is implemented as !IN(left, right)
+		op = In
+	}
+
+	lType := reflect.TypeOf(dummyLeft)
+	rType := reflect.TypeOf(dummyRight)
+
+	// TODO(pmattis): Memoize the cmpOps lookup as we've done for unaryOps and
+	// binOps.
+	if f, ok := cmpOps[cmpArgs{op, lType, rType}]; ok {
```

Comment:
It looks like this will fail to find a comparison op if either argument is `DNull`. I'm not sure that is correct. `a = NULL` is ok from a type-checking perspective, but semantically always `NULL`.